src/auxiliary/mod.rs

```
@@ -0,0 +1,90 @@
+/// * `validator_address` - The address of validator.
+/// * `block` - The observed block which needs to be reported.
+pub fn report_block(
+    event_loop: &tokio_core::reactor::Handle,
+    block_store_contract: &Arc<Contract<Http>>,
+    validator_address: Address,
+    block: &Block,
+) {
+    info!("Reporting block for number {:?} ", block.number);
+
+    let encoded_block = rlp::encode(block);
+    let block_hash = block.hash();
+    let block_store_contract = Arc::clone(&block_store_contract);
+    let call_future = block_store_contract
+        .query(
+            "isBlockReported",
+            block_hash,
+            validator_address,
+            Options::default(),
+            None,
+        ).then(
+            move |result: Result<bool, web3::contract::Error>| match result {
+                Ok(is_reported) => if is_reported {
+                    Either::A(Ok(()).into_future())
+                } else {
+                    Either::B(
+                        block_store_contract
+                            .call(
+                                "reportBlock",
+                                encoded_block,
+                                validator_address.into(),
+                                Options::with(|opt| {
+                                    opt.gas = Some(REPORT_BLOCK_ESTIMATED_GAS.into())
+                                }),
+                            ).then(move |tx| {
+                                info!("Block reported got tx: {:?}", tx);
+                                Ok(())
+                            }),
+                    )
+                },
+                Err(error) => {
+                    error!(
+                        "Error while checking if block is already reported{:?}",
+                        error
+                    );
+                    // Event loop spawn expects certain types. It doesn't support err types.
+                    Either::A(Ok(()).into_future())
```

Also if you return the future and let the caller spawn, you could return an error that the caller could react to.

---

Similar to the above comment, to me reactor single atomic thing. It should also know how to handle the error.